jose-jwt/jwe/JWE.cs

```diff
@@ -0,0 +1,330 @@
+        {
+            if (plaintext == null)
+            {
+                throw new ArgumentNullException(nameof(plaintext));
+            }
+
+            settings = GetSettings(settings);
+            IJweAlgorithm _enc = settings.Jwe(enc);
+
+            if (_enc == null)
+            {
+                throw new JoseException(string.Format("Unsupported JWE enc requested: {0}", enc));
+            }
+
+            IDictionary<string, object> joseProtectedHeader = MergeHeaders(
+                new Dictionary<string, object> { { "enc", settings.JweHeaderValue(enc) } },
+                extraHeaders);
+
+            byte[] cek = null;
+
+            var recipientsOut = new List<(byte[] EncryptedKey, IDictionary<string, object> Header)>();
+            foreach (var recipient in recipients)
+            {
+                IKeyManagement keys = settings.Jwa(recipient.Alg);
+
+                if (keys == null)
+                {
+                    throw new JoseException(string.Format("Unsupported JWE alg requested: {0}", recipient.Alg));
+                }
+
+                // joseHeader - is merge of headers
+                // - key management will read from (e.g. enc,apv,apu - ECDH-ES)
+                // - key management will write to (e.g. iv, tag - AesGcmKW)
+                IDictionary<string, object> joseHeader = MergeHeaders(
+                    joseProtectedHeader,
+                    new Dictionary<string, object> { { "alg", settings.JwaHeaderValue(recipient.Alg) } },
+                    recipient.PerRecipientHeaders);
+
+                byte[] encryptedCek;
+                if(cek == null)
+                {
+                    byte[][] contentKeys = keys.WrapNewKey(_enc.KeySize, recipient.Key, joseHeader);
+                    cek = contentKeys[0];
+                    encryptedCek = contentKeys[1];
+                }
+                else
+                {
+                    encryptedCek = keys.WrapKey(cek, recipient.Key, joseHeader);
+                }
+
+                // For the per-receipient header we want the headers from the result of IKeyManagements key wrapping.. but without the
+                // protected headers that were merged in
+                IDictionary<string, object> recipientHeader = joseHeader.Except(joseProtectedHeader).ToDictionary(x => x.Key, v => v.Value);
+
+                recipientsOut.Add((EncryptedKey: encryptedCek, Header: recipientHeader));
+            }
+
+            if (compression.HasValue)
+            {
+                joseProtectedHeader["zip"] = settings.CompressionHeader(compression.Value);
+                plaintext = settings.Compression(compression.Value).Compress(plaintext);
+            }
+
+            switch (mode)
+            {
+                case SerializationMode.smCompact:
+                    {
+                        if (recipientsOut.Count != 1)

```

Comment:
Compact mode doesn't seem to make anything different from `Jose.JWT.Encode`, right?
What's the benefit to duplicate it here?

Reply:
I've updated PR so that JWT.EncodeBytes now calls through to JWE.Encrypt, so no duplication and JWT class gets simplified.